engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java

```diff
@@ -0,0 +1,216 @@
+	public void not_null_type_on_optional_is_validated_for_null_value() {
+		Model model = new Model();
+		model.valueWithNotNull = null;
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty( model, "valueWithNotNull" );
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithNotNull" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "container" );
+		assertCorrectConstraintTypes( constraintViolations, NotNull.class );
+	}
+
+	/**
+	 * Actual: NotNull violation occurs on container's annotation
+	 */
+	@Test
+	public void not_null_type_on_optional_is_validated_for_empty_value() {
+		Model model = new Model();
+		model.valueWithNotNull = Optional.empty();
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty( model, "valueWithNotNull" );
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithNotNull" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "type" );
+		assertCorrectConstraintTypes( constraintViolations, NotNullTypeUse.class );
+	}
+
+	/**
+	 * Actual: Fails with NullPointerException (Unexpected exception during isValid call)
+	 */
+	@Test
+	public void null_or_not_blank_type_on_optional_is_validated_for_null_value() {
+		Model model = new Model();
+		model.valueWithNullOrNotBlank = null;
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty(
+				model,
+				"valueWithNullOrNotBlank"
+		);
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithNullOrNotBlank" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "container" );
+		assertCorrectConstraintTypes( constraintViolations, NotNull.class );
+	}
+
+	/**
+	 * Actual: NotNull violation occurs on container's annotation
+	 */
+	@Test
+	public void null_or_not_blank_type_on_optional_is_validated_for_empty_value() {
+		Model model = new Model();
+		model.valueWithNullOrNotBlank = Optional.empty();
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty(
+				model,
+				"valueWithNullOrNotBlank"
+		);
+		assertNumberOfViolations( constraintViolations, 0 );
+	}
+
+	/**
+	 * Actual: No violation
+	 */
+	@Test
+	public void null_or_not_blank_type_on_optional_is_validated_for_blank_string() {
+		Model model = new Model();
+		model.valueWithNullOrNotBlank = Optional.of( "" );
+
+		Set<ConstraintViolation<Model>> constraintViolations = validator.validateProperty(
+				model,
+				"valueWithNullOrNotBlank"
+		);
+		assertNumberOfViolations( constraintViolations, 1 );
+		assertCorrectPropertyPaths( constraintViolations, "valueWithNullOrNotBlank" );
+		assertConstraintViolation( constraintViolations.iterator().next(), "type" );
+		assertCorrectConstraintTypes( constraintViolations, NullOrNotBlank.class );
+	}
+
+	@Test
```

Comment:
as far as I understand the issue is about to test the ability of validator to validate the wrapper object and the wrapped object separately.

This test is checking that the validation fails because of the constraint on the type and not the one on the wrapper.

There is another one on the same field that checks that when the container is null, the test fails only for the validation of the container `null_or_not_blank_type_on_optional_is_validated_for_null_value`.